// @react-navigation/compat/lib/module/StackActions.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import { StackActions } from '@react-navigation/native';
export function reset() {
  throw new Error('The legacy `reset` action is not supported. Use the new reset API by accessing the original navigation object at `navigation.original`.');
}
export function replace({
  routeName,
  params,
  key,
  newKey,
  action
}) {
  if (action !== undefined) {
    throw new Error('Sub-actions are not supported for `replace`. Remove the `action` key from the options.');
  }

  return _objectSpread({
    type: 'REPLACE',
    payload: {
      name: routeName,
      key: newKey,
      params
    }
  }, key !== undefined ? {
    source: key
  } : null);
}
export function push({
  routeName,
  params,
  action
}) {
  if (action !== undefined) {
    throw new Error('Sub-actions are not supported for `push`. Remove the `action` key from the options.');
  }

  return StackActions.push(routeName, params);
}
export function pop({
  n = 1
}) {
  return StackActions.pop(n);
}
export function popToTop() {
  return StackActions.popToTop();
}
export function dismiss() {
  throw new Error('The legacy `dismiss` action is not supported.');
}
